packages/appium/lib/appium.js

```diff
@@ -176,23 +176,38 @@ class AppiumDriver extends BaseDriver {
         log.info(`Applying relaxed security to '${InnerDriver.name}' as per ` +
                  `server command line argument. All insecure features will be ` +
                  `enabled unless explicitly disabled by --deny-insecure`);
-        d.relaxedSecurityEnabled = true;
+        driverInstance.relaxedSecurityEnabled = true;
       }
 
       if (!_.isEmpty(this.args.denyInsecure)) {
         log.info('Explicitly preventing use of insecure features:');
         this.args.denyInsecure.map((a) => log.info(`    ${a}`));
-        d.denyInsecure = this.args.denyInsecure;
+        driverInstance.denyInsecure = this.args.denyInsecure;
       }
 
       if (!_.isEmpty(this.args.allowInsecure)) {
         log.info('Explicitly enabling use of insecure features:');
         this.args.allowInsecure.map((a) => log.info(`    ${a}`));
-        d.allowInsecure = this.args.allowInsecure;
+        driverInstance.allowInsecure = this.args.allowInsecure;
+      }
+
+      // Likewise, any driver-specific CLI args that were passed in should be assigned directly to
+      // the driver so that they cannot be mimicked by a malicious user sending in capabilities
+      const driverCliArgs = getExtensionArgs(this.args.driverArgs, driverName);
+      if (!_.isEmpty(driverCliArgs)) {
+        if (!_.has(InnerDriver, 'argsConstraints')) {
+          throw new Error(`You sent in CLI args for the ${driverName} driver, but the driver ` +
+                          `does not define any`);
+        }
+        // parse/verify CLI args and merge them to this.args if OK or throw an exception
+        validateExtensionArgs(driverCliArgs, InnerDriver.argsConstraints);
+        driverInstance.cliArgs = driverCliArgs;
```

Comment:
where do we want to put the logic that reflects cliArgs onto driver capabilities (basically rewrites or adds them if needed)?

Reply:
I don't think we do. Drivers need to be able to distinguish what has been passed in as a CLI arg and what has been passed in as a capability. I don't want someone to be able to send in a capability of the same name as a CLI arg and make the driver think that CLI arg has been used. CLI args are for sysadmins only (like setting the security flags etc).

So that is why we keep cli args on `driver.cliArgs` and capabilities get merged to `driver.opts`.

Reply:
yes, this makes sense. I just wanted to know where exactly capabilities should be modified based on cliArgs. For example: 
There is --chromedriver-executable uia2 arg. what it does is basically adds/overrides the corresponding capability, so the executable is always takes from the corresponding server argument instead of user-provided caps. Now my questions is: we have `cliArgs`, where the value of the arg is stored. Where do we want to override capabilities with this value (and how) : in the uia2 driver code or in the umbrella driver code? I'm asking because we also need to verify caps and the override must happen before the verification.

Reply:
why does the override need to happen before verification? do you think my solution in https://github.com/appium/appium-xcuitest-driver/pull/1331 would work?

Reply:
added my comments to that PR